app/logic/Calendar/Invitation/IncomingInvitation.ts

```diff
@@ -0,0 +1,63 @@
+import type { Calendar } from "../Calendar";
+import type { Event } from "../Event";
+import { InvitationMessage, InvitationResponse, type InvitationResponseInMessage } from "../Invitation/InvitationStatus";
+import type { EMail } from "../../Mail/EMail";
+import { assert } from "../../util/util";
+
+export class IncomingInvitation {
+  readonly calendar: Calendar;
+  readonly message: EMail;
+  readonly invitationMessage: InvitationMessage;
+  readonly event: Event;
+  myParticipation: InvitationResponse;
+
+  constructor(calendar: Calendar, message: EMail) {
+    this.calendar = calendar;
+    this.message = message;
+    this.invitationMessage = message.invitationMessage;
+    this.event = message.event;
+    let event = calendar.events.find(event => event.calUID == this.event.calUID);
+    this.myParticipation = event?.myParticipation || InvitationResponse.NoResponseReceived;
+  }
+
+  async respondToInvitation(response: InvitationResponseInMessage) {
+    assert(this.invitationMessage == InvitationMessage.Invitation, "Only invitations can be responded to");
+    let event = this.calendar.events.find(event => event.calUID == this.event.calUID);
+    if (!event) {
+      event = this.calendar.newEvent();
+      event.copyFrom(this.event);
+      event.recurrenceRule = this.event.recurrenceRule;
+      this.calendar.events.add(event);
+      if (event.recurrenceRule) {
+        event.fillRecurrences();
+      }
+    }
+    let { myParticipant } = event.participantMe(this.message.folder.account);
+    event.myParticipation = myParticipant.response = response;
+    await event.save();
+    await event.sendInvitationResponse(myParticipant, this.message.folder.account);
+    /* else add participant? */
+  }
+
+  /** Handle the other two InvitationMessage cases:
+   *  CancelledEvent: the organiser cancelled an incoming meeting
+   *  ParticpantReply: an invitee replied to your outgoing invitation */
+  async updateFromOtherInvitationMessage() {
+    assert(this.invitationMessage && this.invitationMessage != InvitationMessage.Invitation, "can't update from an invitation");
+    let event = this.calendar.events.find(p => p.calUID == this.event.calUID);
+    if (this.invitationMessage == InvitationMessage.CancelledEvent) {
+      let organizer = event.participants.find(participant => participant.response == InvitationResponse.Organizer);
+      if (organizer) {
+        organizer.response = InvitationResponse.Decline;
+        await event.save();
+      }
+    } else {
+      let invitee = this.event.participants.find(participant => participant.response != InvitationResponse.Organizer);
```

Comment:
You're assuming that there will be only 2 participants in the response: The organizer and the sender. Is that mandated by the spec? If not, it's conceivable that there is some software that sends *all* participants. (Even if you have not actually encountered such a message. We have not seen all software that sends iCal.)

I think there should be an extra check that either a) there are <= 2 participants in the list or b) the participant email address matches the From email address of the email.

Reply:
in theory there is apparently something called delegation - you receive an invitation, but you want someone else to take your place. In this case you're supposed to send the organiser a message that you want to delegate the invitation and the other person sends to both people whether they can attend, and the resulting message has three participants.

Unfortunately whenever I try to search for support for this feature I only get the completely different feature of Office 365/Exchange whereby you can look at another user's calendar, inbox etc.

Reply:
Yes, I don't think Exchange does it this way.

Reply:
I was thinking about the case where some calendar client on the other end simply copies *all* participants from the original invitation that they received, copied all their status, and changed only their own status. I am not sure such a software exists, but it's conceivable, and we should not break in this case.

Reply:
Also, this needs a null check. If there is no other participant is found, `invitee` will be null, and `invitee.emailAddress` and `invitee.response` below will throw.

Reply:
Given that, how should we notify the user that the reply is malformed?

Reply:
I guess the best would be an inline error message inside the invitation display.